# PatchTST — S&P 500 Equity Straddles

GBM leads all families on IC (0.068) for this case study. PatchTST adds temporal structure:
it groups 60 days of IV features into patches and applies self-attention,
testing whether the *evolution* of volatility surfaces — not just their
current snapshot — improves delta-hedged straddle return prediction.

**Learning Objectives**:
- Quantify whether temporal IV dynamics add predictive value over flat features
- Compare patch-based attention to LSTM gating on options data
- Track convergence speed (these 612 symbols converge fast)

**Book Reference**: Chapter 13

**Prerequisites**: [`06_linear`](06_linear.ipynb), [`07_gbm`](07_gbm.ipynb) (for comparison baselines)

In [1]:
"""PatchTST — sp500_options deep learning."""

import warnings

import numpy as np
import polars as pl
import torch
import yaml

from case_studies.utils.analytics import load_best_ic_per_family
from case_studies.utils.deep_learning import (
    create_model,
    resolve_arch_name,
    run_dl_cv,
)
from utils.modeling import load_configs, load_modeling_dataset
from utils.paths import get_case_study_dir

warnings.filterwarnings("ignore")

In [2]:
CASE_STUDY_ID = "sp500_options"
MODEL = "patchtst"
PRIMARY_LABEL = ""
MAX_SYMBOLS = 0
N_EPOCHS = 100
LOOKBACK = 60
BATCH_SIZE = 2048
MC_DROPOUT = False
MAX_FOLDS = 0
FORCE_RETRAIN = False  # Set True to retrain configs that already have complete hashes
PREDICTION_SPLIT = "validation"

In [3]:
CASE_DIR = get_case_study_dir(CASE_STUDY_ID)
setup = yaml.safe_load((CASE_DIR / "config" / "setup.yaml").read_text())

if not PRIMARY_LABEL:
    PRIMARY_LABEL = setup["labels"]["primary"]
    print(f"Label from setup.yaml: {PRIMARY_LABEL}")
else:
    print(f"Label override: {PRIMARY_LABEL}")

dl_config = setup.get("modeling", {}).get("dl", {})
DEVICE = dl_config.get("device", "gpu")

device_str = "cuda" if DEVICE == "gpu" and torch.cuda.is_available() else "cpu"
print(f"Case study: {CASE_STUDY_ID} | Model: {MODEL}")
print(f"Device: {device_str} | Epochs: {N_EPOCHS} | Lookback: {LOOKBACK}")

Label from setup.yaml: fwd_ret_dh_10d
Case study: sp500_options | Model: patchtst
Device: cuda | Epochs: 100 | Lookback: 60


## 1. Load Data

In [4]:
mds = load_modeling_dataset(CASE_STUDY_ID, PRIMARY_LABEL, max_symbols=MAX_SYMBOLS)

dataset = mds.dataset
feature_names = mds.feature_names
label_col = mds.label_col
date_col = mds.date_col
entity_col = mds.entity_cols[0] if mds.entity_cols else "symbol"
splits = mds.splits
if MAX_FOLDS:
    splits = splits[:MAX_FOLDS]
n_features = len(feature_names)

print(f"Dataset: {len(dataset):,} rows × {n_features} features")
print(f"Label: {label_col} | Entity: {entity_col} | Folds: {len(splits)}")

dataset_pd = dataset.to_pandas()
n_entities = dataset_pd[entity_col].nunique()
print(f"Entities: {n_entities}")

Dataset: 325,089 rows × 51 features
Label: fwd_ret_dh_10d | Entity: symbol | Folds: 2
Entities: 610


## 2. Prior Baselines

Load IC results from earlier pipeline stages (Ch11 linear, Ch12 GBM)
rather than re-running them here.

In [5]:
prior_baselines = {}
_baselines = load_best_ic_per_family(["linear", "gbm"], case_studies=[CASE_STUDY_ID])
if not _baselines.is_empty():
    for row in _baselines.iter_rows(named=True):
        if row["family"] == "linear":
            prior_baselines[f"{row['config_name'].title()} (Ch11)"] = row["ic_mean"]
        elif row["family"] == "gbm":
            prior_baselines["GBM (Ch12)"] = row["ic_mean"]

if prior_baselines:
    for name, ic in prior_baselines.items():
        print(f"  {name}: IC={ic:+.4f}" if ic is not None else f"  {name}: IC=N/A")
else:
    print("  No prior results found — run 06_linear.py and 07_gbm.py first")

  GBM (Ch12): IC=+0.0742
  Ridge_A10000.0 (Ch11): IC=+0.0442


## 3. PatchTST

Primary architecture for this notebook.

In [6]:
dl_configs = load_configs(CASE_STUDY_ID, PRIMARY_LABEL, "deep_learning")
dl_configs = [c for c in dl_configs if c["params"].get("architecture") == MODEL]

# Apply Papermill overrides to configs (test mode: fewer epochs)
for cfg in dl_configs:
    if cfg.get("n_epochs", 100) != N_EPOCHS:
        cfg["n_epochs"] = N_EPOCHS
    if cfg.get("batch_size", 2048) != BATCH_SIZE:
        cfg["batch_size"] = BATCH_SIZE
    if cfg["params"].get("lookback", 60) != LOOKBACK:
        cfg["params"]["lookback"] = LOOKBACK

print(
    f"Grid: {len(dl_configs)} configs × {dl_configs[0].get('n_epochs', 100)} epochs × {len(splits)} folds"
)
for cfg in dl_configs:
    print(
        f"  {cfg['config_name']}: {cfg['params'].get('architecture', '?')} ({cfg.get('n_epochs', 100)} epochs)"
    )

Grid: 1 configs × 100 epochs × 2 folds
  patchtst: patchtst (100 epochs)


In [7]:
result = run_dl_cv(
    dataset_pd,
    splits,
    feature_names=feature_names,
    label_col=label_col,
    date_col=date_col,
    entity_col=entity_col,
    configs=dl_configs,
    n_features=n_features,
    device=device_str,
    save_dir=CASE_DIR / "run_log" / "training" / "deep_learning",
    register=True,
    force_retrain=FORCE_RETRAIN,
    case_study=CASE_STUDY_ID,
    notebook=f"dl_{MODEL}",
    temporal_by_fold=mds.temporal_by_fold,
    temporal_keys=mds.temporal_keys,
    temporal_feature_names=mds.temporal_feature_names,
)

Fold-major CV: 2 folds × 1 configs × 60 lookback

  Fold 0: creating sequences...


    train=94,736 seq across 455 symbols
    val=44,253 seq across 410 symbols
    creating datasets...
    datasets ready
    patchtst:


      epoch   1/100: train_loss=0.093889


      epoch   2/100: train_loss=0.047772


      epoch   3/100: train_loss=0.046982


      epoch   4/100: train_loss=0.046265


      epoch   5/100: train_loss=0.045300, val_loss=0.041436, IC=+0.0476


      epoch   6/100: train_loss=0.044214


      epoch   7/100: train_loss=0.043595


      epoch   8/100: train_loss=0.042593


      epoch   9/100: train_loss=0.041212


      epoch  10/100: train_loss=0.040532, val_loss=0.042774, IC=+0.0212


      epoch  11/100: train_loss=0.039029


      epoch  12/100: train_loss=0.037695


      epoch  13/100: train_loss=0.036565


      epoch  14/100: train_loss=0.035534


      epoch  15/100: train_loss=0.034143, val_loss=0.047426, IC=+0.0171


      epoch  16/100: train_loss=0.033049


      epoch  17/100: train_loss=0.031931


      epoch  18/100: train_loss=0.030915


      epoch  19/100: train_loss=0.030234


      epoch  20/100: train_loss=0.029385, val_loss=0.045503, IC=+0.0210


      epoch  21/100: train_loss=0.028579


      epoch  22/100: train_loss=0.028007


      epoch  23/100: train_loss=0.027203


      epoch  24/100: train_loss=0.026504


      epoch  25/100: train_loss=0.026662, val_loss=0.048458, IC=+0.0192


      epoch  26/100: train_loss=0.025424


      epoch  27/100: train_loss=0.025019


      epoch  28/100: train_loss=0.024527


      epoch  29/100: train_loss=0.023879


      epoch  30/100: train_loss=0.023583, val_loss=0.047765, IC=+0.0135


      epoch  31/100: train_loss=0.023221


      epoch  32/100: train_loss=0.022816


      epoch  33/100: train_loss=0.022625


      epoch  34/100: train_loss=0.022550


      epoch  35/100: train_loss=0.021890, val_loss=0.052834, IC=+0.0190


      epoch  36/100: train_loss=0.021514


      epoch  37/100: train_loss=0.021462


      epoch  38/100: train_loss=0.020947


      epoch  39/100: train_loss=0.020517


      epoch  40/100: train_loss=0.020440, val_loss=0.051985, IC=+0.0175


      epoch  41/100: train_loss=0.020184


      epoch  42/100: train_loss=0.020086


      epoch  43/100: train_loss=0.019662


      epoch  44/100: train_loss=0.019457


      epoch  45/100: train_loss=0.019145, val_loss=0.049241, IC=+0.0209


      epoch  46/100: train_loss=0.018969


      epoch  47/100: train_loss=0.018771


      epoch  48/100: train_loss=0.018737


      epoch  49/100: train_loss=0.018456


      epoch  50/100: train_loss=0.018299, val_loss=0.052270, IC=+0.0267


      epoch  51/100: train_loss=0.018226


      epoch  52/100: train_loss=0.017790


      epoch  53/100: train_loss=0.017877


      epoch  54/100: train_loss=0.017465


      epoch  55/100: train_loss=0.017346, val_loss=0.051877, IC=+0.0197


      epoch  56/100: train_loss=0.017352


      epoch  57/100: train_loss=0.017115


      epoch  58/100: train_loss=0.017037


      epoch  59/100: train_loss=0.016790


      epoch  60/100: train_loss=0.016756, val_loss=0.052645, IC=+0.0186


      epoch  61/100: train_loss=0.016627


      epoch  62/100: train_loss=0.016617


      epoch  63/100: train_loss=0.016379


      epoch  64/100: train_loss=0.016331


      epoch  65/100: train_loss=0.016249, val_loss=0.054550, IC=+0.0260


      epoch  66/100: train_loss=0.016160


      epoch  67/100: train_loss=0.016043


      epoch  68/100: train_loss=0.015889


      epoch  69/100: train_loss=0.015840


      epoch  70/100: train_loss=0.015701, val_loss=0.052922, IC=+0.0232


      epoch  71/100: train_loss=0.015658


      epoch  72/100: train_loss=0.015607


      epoch  73/100: train_loss=0.015497


      epoch  74/100: train_loss=0.015302


      epoch  75/100: train_loss=0.015309, val_loss=0.054185, IC=+0.0196


      epoch  76/100: train_loss=0.015324


      epoch  77/100: train_loss=0.015140


      epoch  78/100: train_loss=0.015229


      epoch  79/100: train_loss=0.015131


      epoch  80/100: train_loss=0.015042, val_loss=0.054280, IC=+0.0226


      epoch  81/100: train_loss=0.015003


      epoch  82/100: train_loss=0.014989


      epoch  83/100: train_loss=0.014955


      epoch  84/100: train_loss=0.014891


      epoch  85/100: train_loss=0.014901, val_loss=0.054466, IC=+0.0225


      epoch  86/100: train_loss=0.014894


      epoch  87/100: train_loss=0.014860


      epoch  88/100: train_loss=0.014714


      epoch  89/100: train_loss=0.014825


      epoch  90/100: train_loss=0.014803, val_loss=0.054405, IC=+0.0236


      epoch  91/100: train_loss=0.014754


      epoch  92/100: train_loss=0.014720


      epoch  93/100: train_loss=0.014595


      epoch  94/100: train_loss=0.014520


      epoch  95/100: train_loss=0.014597, val_loss=0.054371, IC=+0.0228


      epoch  96/100: train_loss=0.014688


      epoch  97/100: train_loss=0.014666


      epoch  98/100: train_loss=0.014597


      epoch  99/100: train_loss=0.014608


      epoch 100/100: train_loss=0.014646, val_loss=0.054389, IC=+0.0231


      best_ep=5, IC=+0.0476 (314.8s, 20 checkpoints)



  Fold 1: creating sequences...


    train=112,221 seq across 470 symbols
    val=33,880 seq across 345 symbols
    creating datasets...
    datasets ready
    patchtst:


      epoch   1/100: train_loss=0.112451


      epoch   2/100: train_loss=0.040740


      epoch   3/100: train_loss=0.040027


      epoch   4/100: train_loss=0.039406


      epoch   5/100: train_loss=0.038715, val_loss=0.046760, IC=+0.0283


      epoch   6/100: train_loss=0.037760


      epoch   7/100: train_loss=0.037014


      epoch   8/100: train_loss=0.036018


      epoch   9/100: train_loss=0.035547


      epoch  10/100: train_loss=0.034677, val_loss=0.046740, IC=+0.0619


      epoch  11/100: train_loss=0.033811


      epoch  12/100: train_loss=0.033035


      epoch  13/100: train_loss=0.032080


      epoch  14/100: train_loss=0.031628


      epoch  15/100: train_loss=0.030607, val_loss=0.045486, IC=+0.0930


      epoch  16/100: train_loss=0.029817


      epoch  17/100: train_loss=0.029082


      epoch  18/100: train_loss=0.028143


      epoch  19/100: train_loss=0.027635


      epoch  20/100: train_loss=0.026804, val_loss=0.047634, IC=+0.0855


      epoch  21/100: train_loss=0.026192


      epoch  22/100: train_loss=0.025663


      epoch  23/100: train_loss=0.025133


      epoch  24/100: train_loss=0.024669


      epoch  25/100: train_loss=0.024067, val_loss=0.047944, IC=+0.0864


      epoch  26/100: train_loss=0.023926


      epoch  27/100: train_loss=0.023354


      epoch  28/100: train_loss=0.022886


      epoch  29/100: train_loss=0.022390


      epoch  30/100: train_loss=0.022213, val_loss=0.051571, IC=+0.0802


      epoch  31/100: train_loss=0.021781


      epoch  32/100: train_loss=0.021237


      epoch  33/100: train_loss=0.021257


      epoch  34/100: train_loss=0.021046


      epoch  35/100: train_loss=0.020799, val_loss=0.050344, IC=+0.0681


      epoch  36/100: train_loss=0.020403


      epoch  37/100: train_loss=0.019949


      epoch  38/100: train_loss=0.019746


      epoch  39/100: train_loss=0.019497


      epoch  40/100: train_loss=0.019433, val_loss=0.052637, IC=+0.0744


      epoch  41/100: train_loss=0.019017


      epoch  42/100: train_loss=0.018835


      epoch  43/100: train_loss=0.018633


      epoch  44/100: train_loss=0.018329


      epoch  45/100: train_loss=0.018288, val_loss=0.054252, IC=+0.0711


      epoch  46/100: train_loss=0.018049


      epoch  47/100: train_loss=0.017987


      epoch  48/100: train_loss=0.017829


      epoch  49/100: train_loss=0.017550


      epoch  50/100: train_loss=0.017354, val_loss=0.053948, IC=+0.0697


      epoch  51/100: train_loss=0.017214


      epoch  52/100: train_loss=0.017040


      epoch  53/100: train_loss=0.016995


      epoch  54/100: train_loss=0.016655


      epoch  55/100: train_loss=0.016686, val_loss=0.054509, IC=+0.0669


      epoch  56/100: train_loss=0.016520


      epoch  57/100: train_loss=0.016396


      epoch  58/100: train_loss=0.016206


      epoch  59/100: train_loss=0.016073


      epoch  60/100: train_loss=0.016071, val_loss=0.055748, IC=+0.0717


      epoch  61/100: train_loss=0.015877


      epoch  62/100: train_loss=0.015730


      epoch  63/100: train_loss=0.015669


      epoch  64/100: train_loss=0.015637


      epoch  65/100: train_loss=0.015470, val_loss=0.054967, IC=+0.0642


      epoch  66/100: train_loss=0.015417


      epoch  67/100: train_loss=0.015351


      epoch  68/100: train_loss=0.015166


      epoch  69/100: train_loss=0.015131


      epoch  70/100: train_loss=0.014988, val_loss=0.055551, IC=+0.0623


      epoch  71/100: train_loss=0.014857


      epoch  72/100: train_loss=0.014892


      epoch  73/100: train_loss=0.014809


      epoch  74/100: train_loss=0.014737


      epoch  75/100: train_loss=0.014667, val_loss=0.057092, IC=+0.0652


      epoch  76/100: train_loss=0.014627


      epoch  77/100: train_loss=0.014623


      epoch  78/100: train_loss=0.014558


      epoch  79/100: train_loss=0.014523


      epoch  80/100: train_loss=0.014424, val_loss=0.056483, IC=+0.0646


      epoch  81/100: train_loss=0.014398


      epoch  82/100: train_loss=0.014358


      epoch  83/100: train_loss=0.014291


      epoch  84/100: train_loss=0.014290


      epoch  85/100: train_loss=0.014253, val_loss=0.056558, IC=+0.0605


      epoch  86/100: train_loss=0.014190


      epoch  87/100: train_loss=0.014187


      epoch  88/100: train_loss=0.014124


      epoch  89/100: train_loss=0.014162


      epoch  90/100: train_loss=0.014119, val_loss=0.057271, IC=+0.0608


      epoch  91/100: train_loss=0.014098


      epoch  92/100: train_loss=0.014079


      epoch  93/100: train_loss=0.014165


      epoch  94/100: train_loss=0.014084


      epoch  95/100: train_loss=0.014099, val_loss=0.057017, IC=+0.0612


      epoch  96/100: train_loss=0.013994


      epoch  97/100: train_loss=0.013991


      epoch  98/100: train_loss=0.014054


      epoch  99/100: train_loss=0.014061


      epoch 100/100: train_loss=0.014053, val_loss=0.057233, IC=+0.0619


      best_ep=15, IC=+0.0930 (342.9s, 20 checkpoints)


  patchtst: best_epoch=15, IC=+0.0550 (657.7s)

  Best: patchtst @ epoch 15 (IC=+0.0550)
  Saved to case_studies/sp500_options/run_log/training/deep_learning


## 4. Learning Curves

In [8]:
grid_results = result["grid_results"]
best_name = result["best_config_name"]
best_epoch = result["best_epoch"]
best_ic = result["best_ic"]

curves = result["all_learning_curves"]
if curves.height > 0:
    checkpoints = sorted(curves["epoch"].unique().to_list())
    display_cps = [cp for cp in checkpoints if cp % 20 == 0 or cp == checkpoints[-1]]

    print(f"{'Config':15s}", end="")
    for cp in display_cps:
        print(f" {cp:>7d}", end="")
    print()
    print("-" * (15 + 8 * len(display_cps)))

    for r in grid_results:
        cfg_data = curves.filter(pl.col("config") == r["config_name"])
        print(f"{r['config_name']:15s}", end="")
        for cp in display_cps:
            row = cfg_data.filter(pl.col("epoch") == cp)
            if row.height > 0:
                print(f" {row['ic_mean'][0]:+7.4f}", end="")
            else:
                print(f" {'N/A':>7s}", end="")
        print()

Config               20      40      60      80     100
-------------------------------------------------------
patchtst        +0.0533 +0.0459 +0.0451 +0.0436 +0.0425


## 5. MC Dropout Uncertainty (Optional)

In [9]:
if MC_DROPOUT:
    from ml4t.diagnostic.metrics import cross_sectional_ic

    from case_studies.utils.deep_learning import mc_dropout_predict
    from case_studies.utils.sequence_dataset import (
        materialize_sequences,
        prepare_fold_sequence_stores,
    )

    dates_series = dataset_pd[date_col]
    last_fold = splits[-1]
    train_mask = (dates_series >= last_fold["train_start"]) & (
        dates_series <= last_fold["train_end"]
    )
    val_mask = (dates_series >= last_fold["val_start"]) & (dates_series <= last_fold["val_end"])

    train_store, val_store, _ = prepare_fold_sequence_stores(
        dataset_pd,
        train_mask=train_mask,
        val_mask=val_mask,
        feature_names=feature_names,
        label_col=label_col,
        date_col=date_col,
        entity_col=entity_col,
        lookback=LOOKBACK,
    )
    X_train_seq, y_train_seq, _, _ = materialize_sequences(train_store)
    X_val_seq, y_val_seq, val_dates, val_entities = materialize_sequences(val_store)

    if len(X_train_seq) > 0 and len(X_val_seq) > 0:
        torch_device = torch.device(device_str)
        best_cfg_dict = dl_configs[0]
        arch_name = best_cfg_dict["params"].get(
            "architecture", resolve_arch_name(best_cfg_dict["config_name"])
        )
        from case_studies.utils.deep_learning import build_arch_kwargs

        best_cfg = build_arch_kwargs(
            best_cfg_dict, n_features, best_cfg_dict["params"].get("lookback", 60)
        )
        mc_model = create_model(arch_name, best_cfg).to(torch_device)

        X_t = torch.FloatTensor(X_train_seq).to(torch_device)
        y_t = torch.FloatTensor(y_train_seq).to(torch_device)
        optimizer = torch.optim.AdamW(mc_model.parameters(), lr=1e-3)
        criterion = torch.nn.MSELoss()

        mc_model.train()
        for ep in range(min(N_EPOCHS, 50)):
            idx = torch.randperm(len(X_t))
            for s in range(0, len(X_t), BATCH_SIZE):
                batch = idx[s : s + BATCH_SIZE]
                loss = criterion(mc_model(X_t[batch]), y_t[batch])
                optimizer.zero_grad()
                loss.backward()
                optimizer.step()

        X_v = torch.FloatTensor(X_val_seq).to(torch_device)
        mean_pred, std_pred = mc_dropout_predict(mc_model, X_v, n_samples=50)

        median_unc = np.median(std_pred)
        low_unc = std_pred <= median_unc
        high_unc = std_pred > median_unc

        low_frame = pl.DataFrame(
            {
                "date": val_dates[low_unc],
                "symbol": val_entities[low_unc],
                "y_true": y_val_seq[low_unc],
                "y_pred": mean_pred[low_unc],
            }
        )
        ic_low = cross_sectional_ic(
            low_frame,
            low_frame,
            pred_col="y_pred",
            ret_col="y_true",
            date_col="date",
            entity_col="symbol",
            min_obs=5,
        )["ic_mean"]
        high_frame = pl.DataFrame(
            {
                "date": val_dates[high_unc],
                "symbol": val_entities[high_unc],
                "y_true": y_val_seq[high_unc],
                "y_pred": mean_pred[high_unc],
            }
        )
        ic_high = cross_sectional_ic(
            high_frame,
            high_frame,
            pred_col="y_pred",
            ret_col="y_true",
            date_col="date",
            entity_col="symbol",
            min_obs=5,
        )["ic_mean"]
        print("MC Dropout uncertainty analysis:")
        print(f"  Low uncertainty IC:  {ic_low:+.4f} ({low_unc.sum():,} samples)")
        print(f"  High uncertainty IC: {ic_high:+.4f} ({high_unc.sum():,} samples)")
        print(f"  IC gap: {ic_low - ic_high:+.4f}")

        del mc_model, X_t, y_t, X_v
        torch.cuda.empty_cache()
else:
    print("MC Dropout disabled (set MC_DROPOUT=True to enable)")

MC Dropout disabled (set MC_DROPOUT=True to enable)


## 6. Comparison

In [10]:
rows = [(name, ic) for name, ic in prior_baselines.items()]
rows.append((best_name, best_ic))

comparison = pl.DataFrame({"Model": [r[0] for r in rows], "IC": [r[1] for r in rows]})
comparison = comparison.with_columns(
    pl.when(pl.col("IC") == pl.col("IC").max())
    .then(pl.lit("*"))
    .otherwise(pl.lit(""))
    .alias("Best")
)
comparison

Model,IC,Best
str,f64,str
"""GBM (Ch12)""",0.074176,"""*"""
"""Ridge_A10000.0 (Ch11)""",0.044249,""""""
"""patchtst""",0.055026,""""""


In [11]:
ridge_ic = next((v for k, v in prior_baselines.items() if "ridge" in k.lower()), float("nan"))
dl_delta = best_ic - ridge_ic
print(f"DL delta over Ridge: {dl_delta:+.4f}")

DL delta over Ridge: +nan


## 7. Save Results

Predictions and fold metrics are registered by `run_dl_cv()`
during training. Here we record the pipeline results JSON.

In [12]:
predictions = result["predictions"]
all_predictions = result["all_predictions"]
fold_metrics = result["fold_metrics"]

print(f"Predictions: {predictions.height:,} rows")
print(f"All predictions: {all_predictions.height:,} rows")

Predictions: 78,133 rows
All predictions: 1,562,660 rows


In [13]:
val_ic_mean = float(fold_metrics["ic_mean"].mean()) if fold_metrics.height > 0 else None

## 8. Key Takeaways

1. **PatchTST outperforms the Ridge baseline** by a meaningful margin,
   confirming that temporal IV dynamics add meaningful value to straddle
   return prediction.
2. **Best DL model**: PatchTST (IC ~0.047) outperforms LSTM (~0.037).
   Patch-based attention captures IV surface dynamics better than gating on
   this cross-sectional data.
3. **Fast convergence (epoch 10)**: The dense signal from 612 equity
   straddles with IV-derived features saturates DL capacity early.
4. **All sequence models lag TabM**: The temporal dimension adds meaningful
   IC over Ridge, but TabM's flat-feature interactions add substantially more
   over GBM. For options, cross-feature structure dominates temporal dynamics.